code/drasil-database/lib/Database/Drasil/ChunkDB.hs

```diff
@@ -1,312 +1,274 @@
+insertRefExpectingExistence depdnt depdncy cbu =
+  case M.insertLookupWithKey (\_ _ (c, depdnts) -> (c, depdnt : depdnts)) depdncy (undefined, []) cbu of
+    (Just _, cbu') -> cbu' -- If the chunk is already registered, we just updated its dependants, and everything is fine.
+    (Nothing, _) -> -- But if no data was found, then we have a problem: the chunk we are inserting depends on a chunk that does not exist.
+      error $ "Chunk dependancy is missing for `" ++ show depdnt ++ "`. Missing: `" ++ show depdncy ++ "`."
+
+-- | Internal function to insert a chunk into the 'ChunkDB'. This function
+-- assumes that the chunk is not already registered in the database, and quietly
+-- break table synchronicity if it is.
+insert0 :: IsChunk a => ChunkDB -> a -> ChunkDB
+insert0 cdb c = cdb'
+  where
+    -- Box our chunk.
+    c' = mkChunk c
+
+    -- Insert our chunk, it is not currently depended on by anything.
+    chunkTable' = M.insert (c ^. uid) (c', mempty) (chunkTable cdb)
+
+    -- Capture all dependencies of this chunk.
+    chunkTable'' = foldr (insertRefExpectingExistence $ c' ^. uid) chunkTable'
+      $ chunkRefs c
+
+    -- Add our chunk to its corresponding 'chunks by type' list.
+    chunkTypeTable' = M.alter (Just . maybe [c'] (++ [c'])) (typeOf c) (chunkTypeTable cdb)
+
+    -- Finally, build the updated database.
+    cdb' = cdb { chunkTable = chunkTable'', chunkTypeTable = chunkTypeTable' }
+
+-- | Insert a chunk into the 'ChunkDB' if it is sensibly to do so (i.e., does
+-- not depend on itself and is not a 'ChunkDB'). We temporarily allow chunks to
+-- overwrite other ones, but we warn when this happens.
+insert :: IsChunk a => a -> ChunkDB -> ChunkDB
+insert c cdb
+  | c ^. uid `elem` chunkRefs c =
+      error $ "Chunk `" ++ show (c ^. uid) ++ "` cannot reference itself as a dependancy."
+  | typeOf c == typeRep (Proxy @ChunkDB) =
+      error "Insertion of ChunkDBs in ChunkDBs is disallowed; please perform unions with them instead."
+  | (Just x) <- findTypeOf (c ^. uid) cdb =
+      -- Overwrite: remove previous chunk from chunk refs and chunksByType
+      -- tables before inserting new one.
+      let prevChunk = fst $ fromMaybe (error "Internal error: chunk missing after findTypeOf succeeded") (M.lookup (c ^. uid) (chunkTable cdb))
+          prevType  = chunkType prevChunk
+          cu' = M.delete (c ^. uid) (chunkTable cdb)
+          ctr' = M.adjust (filter (\c_ -> (c_ ^. uid) /= (c ^. uid))) prevType (chunkTypeTable cdb)
+          cdb' = cdb { chunkTable = cu', chunkTypeTable = ctr' }
+          cdb'' = insert0 cdb' c
+      in if typeOf c == x
+            then trace (warnMsg $ "WARNING! Overwriting `" ++ show (c ^. uid) ++ "` :: " ++ show x) cdb''
+            else trace (errMsg $ "SUPER-MEGA-ULTRA-DELUXE-WARNING! Overwriting a chunk (`" ++ show (c ^. uid) ++ "` :: `" ++ show x ++ "`) with a chunk of a different type: `" ++ show (typeOf c) ++ "`") cdb''
```

Comment:
If #4330 is merged before this PR, we can turn the SUPER-MEGA-ULTRA-DELUXE-WARNINGs into hard errors rather than red coloured warnings.

Reply:
And indeed, that has come to pass.

Reply:
Enabled in [dba5fbe](https://github.com/JacquesCarette/Drasil/pull/4320/commits/dba5fbe191e2bc8a9053d6ade1960ed32236d030).